packages/back-end/src/models/VisualChangesetModel.ts

```diff
@@ -222,63 +221,228 @@ export async function updateVisualChange({
+  id: uniqid("vc_"),
+  variation: variation.id,
+  description: "",
+  css: "",
+  domMutations: [],
+});
+
 export const createVisualChangeset = async ({
   experiment,
   organization,
   urlPatterns,
   editorUrl,
 }: {
   experiment: ExperimentInterface;
-  organization: string;
+  organization: OrganizationInterface;
   urlPatterns: VisualChangesetURLPattern[];
   editorUrl: VisualChangesetInterface["editorUrl"];
 }): Promise<VisualChangesetInterface> => {
-  const visualChangeset = await VisualChangesetModel.create({
-    id: uniqid("vcs_"),
-    experiment: experiment.id,
+  const visualChangeset = toInterface(
+    await VisualChangesetModel.create({
+      id: uniqid("vcs_"),
+      experiment: experiment.id,
+      organization: organization.id,
+      urlPatterns,
+      editorUrl,
+      visualChanges: experiment.variations.map(_genNewVisualChange),
+    })
+  );
+  await _onVisualChangesetCreate({
     organization,
-    urlPatterns,
-    editorUrl,
-    visualChanges: experiment.variations.map((variation) => ({
-      id: uniqid("vc_"),
-      variation: variation.id,
-      description: "",
-      css: "",
-      domMutations: [],
-    })),
+    visualChangeset,
+    experiment,
   });
-  return toInterface(visualChangeset);
+  return visualChangeset;
 };
 
+// type guard
+const _isUpdatingVisualChanges = (
+  updates: Partial<VisualChangesetInterface>
+): updates is {
+  visualChanges: VisualChange[];
+} & Partial<VisualChangesetInterface> =>
+  updates.visualChanges !== undefined && updates.visualChanges.length > 0;
+
 export const updateVisualChangeset = async ({
   changesetId,
   organization,
   updates,
 }: {
   changesetId: string;
-  organization: string;
+  organization: OrganizationInterface;
   updates: Partial<VisualChangesetInterface>;
 }) => {
+  const visualChangeset = await findVisualChangesetById(
+    changesetId,
+    organization.id
+  );
+
+  if (!visualChangeset) {
+    throw new Error("Visual Changeset not found");
+  }
+
+  const isUpdatingVisualChanges = _isUpdatingVisualChanges(updates);
+
   // ensure new visual changes have ids assigned
-  const visualChanges =
-    updates.visualChanges?.map((vc) => ({
-      ...vc,
-      id: vc.id || uniqid("vc_"),
-    })) ?? [];
+  const visualChanges = isUpdatingVisualChanges
+    ? updates.visualChanges.map((vc) => ({
+        ...vc,
+        id: vc.id || uniqid("vc_"),
+      }))
+    : [];
 
   const res = await VisualChangesetModel.updateOne(
     {
       id: changesetId,
-      organization,
+      organization: organization.id,
     },
     {
       $set: {
         ...updates,
-        visualChanges,
+        ...(isUpdatingVisualChanges ? { visualChanges } : {}),
       },
     }
   );
 
+  await _onVisualChangesetUpdate({
+    oldVisualChangeset: visualChangeset,
+    newVisualChangeset: {
+      ...visualChangeset,
+      ...updates,
+      ...(isUpdatingVisualChanges ? { visualChanges } : {}),
```

Comment:
totally unimportant, but this can be simplified to:
`...(isUpdatingVisualChanges && { visualChanges })`

Reply:
TIL. However I don't think it reads well ... THe spread operator used on a non-iterable (boolean) gives me the creeps even if it does work by the grace of javascript.